// purescript-psa/output/Psa.Types/index.js
// Generated by purs version 0.12.0
"use strict";
var Control_Applicative = require("../Control.Applicative/index.js");
var Control_Apply = require("../Control.Apply/index.js");
var Control_Bind = require("../Control.Bind/index.js");
var Control_Monad_ST_Internal = require("../Control.Monad.ST.Internal/index.js");
var Data_Argonaut_Core = require("../Data.Argonaut.Core/index.js");
var Data_Argonaut_Decode = require("../Data.Argonaut.Decode/index.js");
var Data_Argonaut_Decode_Class = require("../Data.Argonaut.Decode.Class/index.js");
var Data_Argonaut_Decode_Combinators = require("../Data.Argonaut.Decode.Combinators/index.js");
var Data_Argonaut_Encode = require("../Data.Argonaut.Encode/index.js");
var Data_Argonaut_Encode_Class = require("../Data.Argonaut.Encode.Class/index.js");
var Data_Either = require("../Data.Either/index.js");
var Data_Eq = require("../Data.Eq/index.js");
var Data_Function = require("../Data.Function/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_Ord = require("../Data.Ord/index.js");
var Data_Ordering = require("../Data.Ordering/index.js");
var Data_Set = require("../Data.Set/index.js");
var Data_Traversable = require("../Data.Traversable/index.js");
var Data_Tuple = require("../Data.Tuple/index.js");
var Foreign_Object = require("../Foreign.Object/index.js");
var Foreign_Object_ST = require("../Foreign.Object.ST/index.js");
var Prelude = require("../Prelude/index.js");
var Unsafe_Coerce = require("../Unsafe.Coerce/index.js");
var NoStats = (function () {
    function NoStats() {

    };
    NoStats.value = new NoStats();
    return NoStats;
})();
var CompactStats = (function () {
    function CompactStats() {

    };
    CompactStats.value = new CompactStats();
    return CompactStats;
})();
var VerboseStats = (function () {
    function VerboseStats() {

    };
    VerboseStats.value = new VerboseStats();
    return VerboseStats;
})();

// | Relative files paths from the cwd, tagged as either being part of the
// | source files or library files of a project. The `Unknown` variant exists
// | because some psc errors are inter-module and aren't reported with a
// | canonical file.
var Src = (function () {
    function Src(value0) {
        this.value0 = value0;
    };
    Src.create = function (value0) {
        return new Src(value0);
    };
    return Src;
})();

// | Relative files paths from the cwd, tagged as either being part of the
// | source files or library files of a project. The `Unknown` variant exists
// | because some psc errors are inter-module and aren't reported with a
// | canonical file.
var Lib = (function () {
    function Lib(value0) {
        this.value0 = value0;
    };
    Lib.create = function (value0) {
        return new Lib(value0);
    };
    return Lib;
})();

// | Relative files paths from the cwd, tagged as either being part of the
// | source files or library files of a project. The `Unknown` variant exists
// | because some psc errors are inter-module and aren't reported with a
// | canonical file.
var Unknown = (function () {
    function Unknown() {

    };
    Unknown.value = new Unknown();
    return Unknown;
})();
var parsePosition = Data_Maybe.maybe(Control_Applicative.pure(Data_Either.applicativeEither)(Data_Maybe.Nothing.value))(function (obj) {
    return Data_Functor.map(Data_Either.functorEither)(Data_Maybe.Just.create)(Control_Apply.apply(Data_Either.applyEither)(Control_Apply.apply(Data_Either.applyEither)(Control_Apply.apply(Data_Either.applyEither)(Data_Functor.map(Data_Either.functorEither)(function (v) {
        return function (v1) {
            return function (v2) {
                return function (v3) {
                    return {
                        startLine: v,
                        startColumn: v1,
                        endLine: v2,
                        endColumn: v3
                    };
                };
            };
        };
    })(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonInt)(obj)("startLine")))(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonInt)(obj)("startColumn")))(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonInt)(obj)("endLine")))(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonInt)(obj)("endColumn")));
});
var maybeProp = function (dictDecodeJson) {
    return function (obj) {
        return function (key) {
            return Data_Maybe.maybe(new Data_Either.Right(Data_Maybe.Nothing.value))(Data_Argonaut_Decode_Class.decodeJson(Data_Argonaut_Decode_Class.decodeJsonMaybe(dictDecodeJson)))(Foreign_Object.lookup(key)(obj));
        };
    };
};
var parseSuggestion = Data_Maybe.maybe(Control_Applicative.pure(Data_Either.applicativeEither)(Data_Maybe.Nothing.value))(function (obj) {
    return Data_Functor.map(Data_Either.functorEither)(Data_Maybe.Just.create)(Control_Apply.apply(Data_Either.applyEither)(Data_Functor.map(Data_Either.functorEither)(function (v) {
        return function (v1) {
            return {
                replacement: v,
                replaceRange: v1
            };
        };
    })(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonString)(obj)("replacement")))(Control_Bind.bind(Data_Either.bindEither)(maybeProp(Data_Argonaut_Decode_Class.decodeForeignObject(Data_Argonaut_Decode_Class.decodeJsonJson))(obj)("replaceRange"))(parsePosition)));
});
var parsePsaError = function (obj) {
    return Control_Apply.apply(Data_Either.applyEither)(Control_Apply.apply(Data_Either.applyEither)(Control_Apply.apply(Data_Either.applyEither)(Control_Apply.apply(Data_Either.applyEither)(Control_Apply.apply(Data_Either.applyEither)(Control_Apply.apply(Data_Either.applyEither)(Data_Functor.map(Data_Either.functorEither)(function (v) {
        return function (v1) {
            return function (v2) {
                return function (v3) {
                    return function (v4) {
                        return function (v5) {
                            return function (v6) {
                                return {
                                    moduleName: v,
                                    errorCode: v1,
                                    errorLink: v2,
                                    message: v3,
                                    filename: v4,
                                    position: v5,
                                    suggestion: v6
                                };
                            };
                        };
                    };
                };
            };
        };
    })(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonMaybe(Data_Argonaut_Decode_Class.decodeJsonString))(obj)("moduleName")))(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonString)(obj)("errorCode")))(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonString)(obj)("errorLink")))(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonString)(obj)("message")))(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonMaybe(Data_Argonaut_Decode_Class.decodeJsonString))(obj)("filename")))(Control_Bind.bind(Data_Either.bindEither)(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonMaybe(Data_Argonaut_Decode_Class.decodeForeignObject(Data_Argonaut_Decode_Class.decodeJsonJson)))(obj)("position"))(parsePosition)))(Control_Bind.bind(Data_Either.bindEither)(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonMaybe(Data_Argonaut_Decode_Class.decodeForeignObject(Data_Argonaut_Decode_Class.decodeJsonJson)))(obj)("suggestion"))(parseSuggestion));
};
var parsePsaResult = function (obj) {
    return Control_Apply.apply(Data_Either.applyEither)(Data_Functor.map(Data_Either.functorEither)(function (v) {
        return function (v1) {
            return {
                warnings: v,
                errors: v1
            };
        };
    })(Control_Bind.bind(Data_Either.bindEither)(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeArray(Data_Argonaut_Decode_Class.decodeForeignObject(Data_Argonaut_Decode_Class.decodeJsonJson)))(obj)("warnings"))(Data_Traversable.traverse(Data_Traversable.traversableArray)(Data_Either.applicativeEither)(parsePsaError))))(Control_Bind.bind(Data_Either.bindEither)(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeArray(Data_Argonaut_Decode_Class.decodeForeignObject(Data_Argonaut_Decode_Class.decodeJsonJson)))(obj)("errors"))(Data_Traversable.traverse(Data_Traversable.traversableArray)(Data_Either.applicativeEither)(parsePsaError)));
};
var eqPsaPath = new Data_Eq.Eq(function (v) {
    return function (v1) {
        if (v instanceof Src && v1 instanceof Src) {
            return v.value0 === v1.value0;
        };
        if (v instanceof Lib && v1 instanceof Lib) {
            return v.value0 === v1.value0;
        };
        if (v instanceof Unknown && v1 instanceof Unknown) {
            return true;
        };
        return false;
    };
});
var ordPsaPath = new Data_Ord.Ord(function () {
    return eqPsaPath;
}, function (v) {
    return function (v1) {
        if (v instanceof Src && v1 instanceof Src) {
            return Data_Ord.compare(Data_Ord.ordString)(v.value0)(v1.value0);
        };
        if (v instanceof Src && v1 instanceof Lib) {
            return Data_Ordering.GT.value;
        };
        if (v instanceof Src && v1 instanceof Unknown) {
            return Data_Ordering.GT.value;
        };
        if (v instanceof Lib && v1 instanceof Src) {
            return Data_Ordering.LT.value;
        };
        if (v instanceof Lib && v1 instanceof Lib) {
            return Data_Ord.compare(Data_Ord.ordString)(v.value0)(v1.value0);
        };
        if (v instanceof Lib && v1 instanceof Unknown) {
            return Data_Ordering.GT.value;
        };
        if (v instanceof Unknown && v1 instanceof Unknown) {
            return Data_Ordering.EQ.value;
        };
        if (v instanceof Unknown) {
            return Data_Ordering.LT.value;
        };
        throw new Error("Failed pattern match at Psa.Types line 58, column 1 - line 58, column 35: " + [ v.constructor.name, v1.constructor.name ]);
    };
});
var encodePosition = Unsafe_Coerce.unsafeCoerce;
var encodeSuggestion = function (suggestion) {
    return Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeForeignObject(Data_Argonaut_Encode_Class.encodeJsonJson))(Foreign_Object.runST(Control_Bind.bind(Control_Monad_ST_Internal.bindST)(Foreign_Object_ST["new"])(function (v) {
        return Control_Bind.bind(Control_Monad_ST_Internal.bindST)(Foreign_Object_ST.poke("replacement")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonJString)(suggestion.replacement))(v))(function (v1) {
            return Control_Bind.bind(Control_Monad_ST_Internal.bindST)(Foreign_Object_ST.poke("replaceRange")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonJson)(Data_Maybe.maybe(Data_Argonaut_Core.jsonNull)(encodePosition)(suggestion.replaceRange)))(v))(function (v2) {
                return Control_Applicative.pure(Control_Monad_ST_Internal.applicativeST)(v);
            });
        });
    })));
};
var encodePsaError = function (error) {
    return Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeForeignObject(Data_Argonaut_Encode_Class.encodeJsonJson))(Foreign_Object.runST(Control_Bind.bind(Control_Monad_ST_Internal.bindST)(Foreign_Object_ST["new"])(function (v) {
        return Control_Bind.bind(Control_Monad_ST_Internal.bindST)(Foreign_Object_ST.poke("moduleName")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonMaybe(Data_Argonaut_Encode_Class.encodeJsonJString))(error.moduleName))(v))(function (v1) {
            return Control_Bind.bind(Control_Monad_ST_Internal.bindST)(Foreign_Object_ST.poke("errorCode")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonJString)(error.errorCode))(v))(function (v2) {
                return Control_Bind.bind(Control_Monad_ST_Internal.bindST)(Foreign_Object_ST.poke("errorLink")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonJString)(error.errorLink))(v))(function (v3) {
                    return Control_Bind.bind(Control_Monad_ST_Internal.bindST)(Foreign_Object_ST.poke("message")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonJString)(error.message))(v))(function (v4) {
                        return Control_Bind.bind(Control_Monad_ST_Internal.bindST)(Foreign_Object_ST.poke("filename")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonMaybe(Data_Argonaut_Encode_Class.encodeJsonJString))(error.filename))(v))(function (v5) {
                            return Control_Bind.bind(Control_Monad_ST_Internal.bindST)(Foreign_Object_ST.poke("position")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonJson)(Data_Maybe.maybe(Data_Argonaut_Core.jsonNull)(encodePosition)(error.position)))(v))(function (v6) {
                                return Control_Bind.bind(Control_Monad_ST_Internal.bindST)(Foreign_Object_ST.poke("suggestion")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonJson)(Data_Maybe.maybe(Data_Argonaut_Core.jsonNull)(encodeSuggestion)(error.suggestion)))(v))(function (v7) {
                                    return Control_Applicative.pure(Control_Monad_ST_Internal.applicativeST)(v);
                                });
                            });
                        });
                    });
                });
            });
        });
    })));
};
var encodePsaResult = function (res) {
    return Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeForeignObject(Data_Argonaut_Encode_Class.encodeJsonJson))(Foreign_Object.runST(Control_Bind.bind(Control_Monad_ST_Internal.bindST)(Foreign_Object_ST["new"])(function (v) {
        return Control_Bind.bind(Control_Monad_ST_Internal.bindST)(Foreign_Object_ST.poke("warnings")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonArray(Data_Argonaut_Encode_Class.encodeJsonJson))(Data_Functor.map(Data_Functor.functorArray)(encodePsaError)(res.warnings)))(v))(function (v1) {
            return Control_Bind.bind(Control_Monad_ST_Internal.bindST)(Foreign_Object_ST.poke("errors")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonArray(Data_Argonaut_Encode_Class.encodeJsonJson))(Data_Functor.map(Data_Functor.functorArray)(encodePsaError)(res.errors)))(v))(function (v2) {
                return Control_Applicative.pure(Control_Monad_ST_Internal.applicativeST)(v);
            });
        });
    })));
};
var compareByLocation = function (err1) {
    return function (err2) {
        var v = Data_Ord.compare(ordPsaPath)(err1.path)(err2.path);
        if (v instanceof Data_Ordering.EQ) {
            if (err1.position instanceof Data_Maybe.Nothing && err2.position instanceof Data_Maybe.Nothing) {
                return Data_Ordering.EQ.value;
            };
            if (err1.position instanceof Data_Maybe.Nothing) {
                return Data_Ordering.LT.value;
            };
            if (err2.position instanceof Data_Maybe.Nothing) {
                return Data_Ordering.GT.value;
            };
            if (err1.position instanceof Data_Maybe.Just && err2.position instanceof Data_Maybe.Just) {
                return Data_Ord.compare(Data_Tuple.ordTuple(Data_Ord.ordInt)(Data_Ord.ordInt))(new Data_Tuple.Tuple(err1.position.value0.startLine, err1.position.value0.startColumn))(new Data_Tuple.Tuple(err2.position.value0.startLine, err2.position.value0.startColumn));
            };
            throw new Error("Failed pattern match at Psa.Types line 120, column 7 - line 126, column 52: " + [ err1.position.constructor.name, err2.position.constructor.name ]);
        };
        return v;
    };
};
module.exports = {
    NoStats: NoStats,
    CompactStats: CompactStats,
    VerboseStats: VerboseStats,
    Src: Src,
    Lib: Lib,
    Unknown: Unknown,
    parsePsaResult: parsePsaResult,
    parsePsaError: parsePsaError,
    encodePsaResult: encodePsaResult,
    encodePsaError: encodePsaError,
    compareByLocation: compareByLocation,
    eqPsaPath: eqPsaPath,
    ordPsaPath: ordPsaPath
};
